### PR TITLE
ETX-500: ensure types are always looked up via package id, not package name

### PR DESCRIPTION
**Pull request by raphael-speyer-da**
tl;dr: The types exist in packages which are indexed by id, so we need to use package ids when doing type lookups.

We load type information from packages, which internally identify themselves by ids.
When a request comes in, it will specify a template id, and when decoding we internally use that id to look up the relevant type in the appropriate package, to know how to decode the JSON.

However, now template ids can use a package name in as their package specifier instead of a package id, so we need to ensure that when doing the type lookups, we are always using the package id to find the relevant package.

Concretely, this involves changing the keys for those lookups from accepting a `ContractTypeId[Ref.PackageRef]` to a `ContractTypeId[Ref.PackageId]`. 

run-all-tests: true

### CODE REVIEW

File: sdk/ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala [hunk 1]
```diff
@@ -352,7 +347,7 @@ object PackageService {
   }
 
   type ResolveTemplateRecordType =
-    ContractTypeId.Template.RequiredPkg => Error \/ typesig.Type
+    ContractTypeId.Template.RequiredPkgId => Error \/ typesig.Type
```
Comment (raphael-speyer-da) [inline sdk/ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala:350]: After substituting these type aliases, we are changing the argument used to lookup a template record type from
`ContractTypeId.Template[Ref.PackageRef]`
to
`ContractTypeId.Template.[Ref.PackageId]`.

This precludes the possibility of specifying the package with a `Ref.PackageName`, which would fail to find anything.

File: sdk/ledger-service/http-json/src/main/scala/com/digitalasset/http/json/DomainJsonDecoder.scala [hunk 2]
```diff
@@ -194,70 +200,79 @@ class DomainJsonDecoder(
...
       ec: ExecutionContext,
       lc: LoggingContextOf[InstanceUUID],
-  ): ET[domain.ContractTypeId.ResolvedOf[CtId]] =
+      overload: PackageService.ResolveContractTypeId.Overload[U, CtId],
+  ): ET[domain.ContractTypeRef[CtId]] =
     eitherT(
       resolveContractTypeId(jwt, ledgerId)(id)
         .map(
           _.toOption.flatten
-            .map(_.original.asInstanceOf[ContractTypeId.ResolvedOf[CtId]])
```
Comment (raphael-speyer-da) [inline sdk/ledger-service/http-json/src/main/scala/com/digitalasset/http/json/DomainJsonDecoder.scala:252]: Previously `templateId_` would internally determine the `ContractTypeRef` for the input id, and then would always return the `original` field, which was a resolved version of it.

The change is to return the `ContractTypeRef` itself here, and let the caller determined where they want to use the `original` or `latestPkgId` field. The `latestPkgId` field must be used for all type lookups, but the `original` may be used for representing the resolved templateId in downstream types.

File: sdk/ledger-service/http-json/src/main/scala/com/digitalasset/http/json/DomainJsonDecoder.scala [hunk 3]
```diff
@@ -194,70 +200,79 @@ class DomainJsonDecoder(
...
   ](
-      id: CtId[Ref.PackageRef],
+      id: U with ContractTypeId.RequiredPkg,
       jwt: Jwt,
       ledgerId: LedgerApiDomain.LedgerId,
   )(implicit
       ec: ExecutionContext,
       lc: LoggingContextOf[InstanceUUID],
-  ): ET[domain.ContractTypeId.ResolvedOf[CtId]] =
+      overload: PackageService.ResolveContractTypeId.Overload[U, CtId],
```
Comment (raphael-speyer-da) [inline sdk/ledger-service/http-json/src/main/scala/com/digitalasset/http/json/DomainJsonDecoder.scala:262]: We re-introduce the implicit overload instead of the `asInstanceOf` below. The purpose is because the type of the `id` argument could be `ContractTypeId.Unknown`, i.e. we don't know whether it's intended to be a template or interface. However the output will always be in reference to a `ContractTypeId.Definite`, i.e. it does know whether we're talking about a template or interface. Because `Unknown`'s can't appear in the output, we can't use the input type to direclty determine the output type, so the implicit `overload` arg provides a mechanism to map from input types to output types.

**Review by erwin-ramirez-da — APPROVED** [review]
LGTM
